File: src/main/java/net/moddingplayground/onastick/api/item/OnAStickItem.java
package net.moddingplayground.onastick.api.item;

import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.screen.NamedScreenHandlerFactory;
import net.minecraft.screen.ScreenHandlerFactory;
import net.minecraft.screen.SimpleNamedScreenHandlerFactory;
import net.minecraft.tag.TagKey;
import net.minecraft.text.Text;
import net.minecraft.util.Hand;
import net.minecraft.util.TypedActionResult;
import net.minecraft.world.World;

public class OnAStickItem extends Item {
    private final ScreenHandlerFactory baseFactory;
    private final Text title;
    private final TagKey<Item> tag;

    public OnAStickItem(Settings settings, ScreenHandlerFactory baseFactory, Text title, TagKey<Item> tag) {
        super(settings);
        this.baseFactory = baseFactory;
        this.title = title;
        this.tag = tag;
    }

    public TagKey<Item> getTag() {
        return this.tag;
    }

    @Override
    public TypedActionResult<ItemStack> use(World world, PlayerEntity user, Hand hand) {
        user.openHandledScreen(this.createScreenHandlerFactory());
        return TypedActionResult.success(user.getStackInHand(hand));
    }

    public NamedScreenHandlerFactory createScreenHandlerFactory() {
        return new SimpleNamedScreenHandlerFactory(this.baseFactory, this.title);
    }
}

File: src/main/kotlin/dev/andante/onastick/item/OnAStickItem.kt
package dev.andante.onastick.item

import dev.andante.onastick.screen.LocalScreenHandlerFactory
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.item.Item
import net.minecraft.item.ItemStack
import net.minecraft.screen.SimpleNamedScreenHandlerFactory
import net.minecraft.text.Text
import net.minecraft.util.Hand
import net.minecraft.util.TypedActionResult
import net.minecraft.world.World

/**
 * Represents an item classified as 'On A Stick'.
 */
class OnAStickItem(
    settings: Settings,

    /**
     * The base screen handler factory.
     */
    baseFactory: LocalScreenHandlerFactory,

    /**
     * The title of the stick's opened screen.
     */
    private val title: Text
) : Item(settings) {
    private val baseFactory = LocalScreenHandlerFactory.build(baseFactory) { this }

    /**
     * Creates a screen handler factory for this stick.
     */
    private val screenHandlerFactory get() =
        SimpleNamedScreenHandlerFactory(baseFactory, title)

    override fun use(world: World, user: PlayerEntity, hand: Hand): TypedActionResult<ItemStack> {
        user.openHandledScreen(screenHandlerFactory)
        return TypedActionResult.success(user.getStackInHand(hand))
    }
}
